src/compiler/crystal/codegen/exception.cr

```diff
@@ -182,6 +183,21 @@ class Crystal::CodeGenVisitor
           # If the rescue restriction matches, codegen the rescue block.
           position_at_end this_rescue_block
 
+          # On windows, we are "inside" the catchpad block. It's difficult to track when to catchret when
+          # codegenning the entire rescue body, so we catchret early and execute the rescue bodies "outside" the
+          # rescue block.
+          if catch_pad = @catch_pad
+            catch_ret_target_block = new_block "this_rescue_target"
+            builder.build_catch_ret catch_pad, catch_ret_target_block
+            position_at_end catch_ret_target_block
+          end
+
+          saved_catch_pad = @catch_pad
+          @catch_pad = old_catch_pad
+
+          # We are generating code for rescues, so set up the rescue block.
+          @rescue_block = rescue_ensure_block
```

Comment:
why there is a need to set the `rescue_block` in every codegen of the node_rescues instead of setting the `@rescue_block` only once as before?

Reply:
`match_any_type_id` can generate a call, and in that case on windows, the call would be before we catchret, but it would be inside the `rescue_ensure_block` so would generate an invoke. This would mean we'd have invokes to the same catchswitch from inside and outside the catchpad which is disallowed. All this moving around is simply to prevent the call to `~match<Type>` from being inside the `rescue_ensure_block`.